### ISSUE TITLE
Dead Locks Happen in lock() Process

### ISSUE THREAD

**Issue by kajelas** [mrniko/redisson#491]
when heavy concurrency happens in my application, a few lock requests will "sink" without any responses, even after the lock lease time has passed. All of these requests wait at `RedissonLock.lockInterruptibly()`.The exact position is `RedissonLock.get()` after `RedissonLock.subscribe()`. 

In my opinion, this may be due to a thread removes the netty listener which is used by another thread. It can happen in this way: 
1. Thread A is in the loop of getting the lock after subscription.
2. Thread B has also applied subscription and waits for result.
3. Thread A gets the lock very soon and enters  `RedissonLock.unsubscribe()`. In this step, it possibly removes all the listeners on the same channel, which includes the listener used by Thread B. It causes Thread B can never get subscription response and hang on forever.

The similar issue is at [https://github.com/mrniko/redisson/pull/93]. But I think it is not solved completely.

Also I suggest  to apply the ttl algorithm to `RedissonLock.get()` because this step can cost some time. And if it has a timeout, dead lock can be prevented in a work-around way.

This is the thread dump when dead lock happens:
`"http-nio-8001-exec-567" #5935 daemon prio=5 os_prio=0 tid=0x00007f15f81ca000 nid=0x6b3e waiting on condition [0x00007f154468a000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000000bf468fc8> (a java.util.concurrent.CountDownLatch$Sync)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(AbstractQueuedSynchronizer.java:836)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.doAcquireSharedInterruptibly(AbstractQueuedSynchronizer.java:997)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireSharedInterruptibly(AbstractQueuedSynchronizer.java:1304)
	at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)
	at org.redisson.command.CommandAsyncService.get(CommandAsyncService.java:95)
	at org.redisson.RedissonObject.get(RedissonObject.java:55)
	at org.redisson.RedissonLock.lockInterruptibly(RedissonLock.java:113)
	at org.redisson.RedissonLock.lock(RedissonLock.java:92)
        ...`

**Comment by mrniko** [comment]
> In my opinion, this may be due to a thread removes the netty listener which is used by another thread

This is not possible because of both `subscribe` and `unsubscribe` synchronized on `PublishSubscribe` object.

> Also I suggest to apply the ttl algorithm to RedissonLock.get() because this step can cost some time

Please, use `RLock.lockInterruptibly(leaseTime, unit)` for this.

**Comment by kajelas** [comment]
@mrniko Thx for your quick reply.

>  Please check your lock usage logic inside your application.

Our logic is very simple as below:

````
//create client as singleton
RedissonClient client = Redisson.create(redissonConfig);

//get lock when a thread needs distribute global locking
RLock lock = client.getLock(lockName);
//locking it
lock.lock(leaseTime,TimeUnit.MILLISECONDS);

//business logic.....

//lock.unlock();
````
Are there any problems?

......................................................
> This is not possible because of both `subscribe` and `unsubscribe` methods are synchronized on `PublishSubscribe` object.

I think it is possible because the while loop in `RedissonLock.lockInterruptibly()` is not synchronized on `PublishSubscribe` object.  Please consider the example I mentioned:

> 1. Thread A is in the loop of getting the lock after subscription.
2. Thread B has also applied subscription and waits for result.
3. Thread A gets the lock very soon and enters RedissonLock.unsubscribe()`. In this step, it possibly removes all the listeners on the same channel, which includes the listener used by Thread B. It causes Thread B can never get subscription response and hang on forever.

......................................................
> Use `RLock.lockInterruptibly(leaseTime, unit)` for this.

Sorry, I can't understand why this method can work. In my understanding, this method can be used to continue the main thread if the async thread is interrupted. But our situation is that `RedissonLock.get()` can not get response, namely it is not related to interruption of async thread. Could you please give some explanation about why this method can work? Thx:)

**Comment by mrniko** [comment]
> I think it is possible because the while loop in RedissonLock.lockInterruptibly() is not synchronized on PublishSubscribe object

Please take a look inside `PublishSubscribe.subscribe` and `PublishSubscribe.unsubscribe` methods

**Comment by mrniko** [comment]
> Are there any problems?

I see `lock.unlock();` is commented out.

**Comment by kajelas** [comment]
> I see `lock.unlock();` is commented out.

It is my mistake. It should not be commented out. So there may be no unsuitable usage of lock in my code, right?

......................................................

> Please take a look inside `PublishSubscribe.subscribe` and `PublishSubscribe.unsubscribe` methods used by `RedissonLock.lockInterruptibly`.

I have already look at these two methods. It seems you mistake what I mean --- **there are no problems in `PublishSubscribe.subscribe` and `PublishSubscribe.unsubscribe`**. The problem may sit in the while loop of `RedissonLock.lockInterruptibly`, which keeps a thread to wait `leaseTime` before failed to get the lock. I wonder maybe this while loop causes Thread A removing listeners used by the Thread B when Thread A is unsubscribing:
> 1. Thread A is in the loop of getting the lock after subscription.
2. Thread B has also applied subscription and waits for result.
3. Thread A gets the lock very soon and enters RedissonLock.unsubscribe()`. In this step, it possibly removes all the listeners on the same channel, which includes the listener used by Thread B. It causes Thread B can never get subscription response and hang on forever.

......................................................

>My mistake. Wait timeout is available only in tryLock(long waitTime, long leaseTime, TimeUnit unit) method, try it.

I love this method:) Very possibly it can solve my problem. But please consider the potential bugs I have mentioned above.

**Comment by mrniko** [comment]
> 1. Thread A is in the loop of getting the lock after subscription.
> 2. Thread B has also applied subscription and waits for result.
> 3. Thread A gets the lock very soon and enters RedissonLock.unsubscribe(). In this step, it possibly removes all the listeners on the same channel, which includes the listener used by Thread B. It causes Thread B can never get subscription response and hang on forever.

Subscribe and unsubscribe methods are based on `acquire/release` approach which increments and decrements object acquisition counter. Then any thread invokes `Subscribe` method counter increments and `unsubscribe` decrements and these operations are synchronized on single object. So it's not possible to `unsubscribe` if acquisition is not equals to zero.

> it possibly removes all the listeners on the same channel, which includes the listener used by Thread B

So this situation is not possible.

**Comment by kajelas** [comment]
Thx a lot. You're correct, the situation I presume is not possible. But why a lot of threads **get stuck at the latch waiting for successful subscription message**:
>"http-nio-8001-exec-567" #5935 daemon prio=5 os_prio=0 tid=0x00007f15f81ca000 nid=0x6b3e waiting on condition [0x00007f154468a000] java.lang.Thread.State: WAITING (parking)
...
> `at java.util.concurrent.CountDownLatch.await(CountDownLatch.java:231)`

>`at org.redisson.command.CommandAsyncService.get(CommandAsyncService.java:95)`
 
>`at org.redisson.RedissonObject.get(RedissonObject.java:55)` 

>`at org.redisson.RedissonLock.lockInterruptibly(RedissonLock.java:113) `
....
My code is simple. It works at a low concurreny correctly but fail in high concurrent situation. What are possbile reasons of such dead lock?

**Comment by mrniko** [comment]
Do you create Redisson instance each time?

**Comment by kajelas** [comment]
Do you mean RedissonClient? `Redisson.create(redissonConfig)` is only invoked at the time when the application starts. All threads share the same RedissonClient instance. Is it not thread-safe?

**Comment by mrniko** [comment]
I'm just checked. Of course it's fully thread-safe as all objects it exposes.

**Comment by kajelas** [comment]
So what could be the causes? I've also checked that our pubsub on the Redis Server works OK.

**Comment by mrniko** [comment]
Have you tried to increase SubscriptionConnectionPoolSize ?

**Comment by kajelas** [comment]
I've considered this config. But we use sentinal mode, and it only has `slaveSubscriptionConnectionPoolSize`. Could this param work? We do not directly connect to Redis Slave

**Comment by mrniko** [comment]
> slaveSubscriptionConnectionPoolSize. Could this param work

Yes. Try it

**Comment by kajelas** [comment]
OK, I'll try it.

By the way, is there any **fair** distributed locking mechanism

**Comment by mrniko** [comment]
@kajelas 

Thanks to current lock architecture it's not hard to implement it. I'll add this feature in future.

**Comment by alexs20** [comment]
Hi,
I can confirm the deadlock in RLock.lock(). I created a simple JUnit class where 20 threads dooing some simple logic like get lock, bucket, semafore. And in 10 runs it periodically stucks on RLock.lock() method. I am using latest redisson library 2.2.14.

**Comment by alexs20** [comment]
Here is the JUnit class.
`package tmp;

import java.util.Arrays;
import java.util.Collection;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.MethodSorters;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import org.redisson.Config;
import org.redisson.Redisson;
import org.redisson.RedissonClient;
import org.redisson.core.RBucket;
import org.redisson.core.RLock;
import org.redisson.core.RSemaphore;

@FixMethodOrder(MethodSorters.NAME_ASCENDING)
@RunWith(Parameterized.class)
public class AVMRedisMTPTest {
  @Parameters
  public static Collection<Object[]> data() {

    return Arrays.asList(new Object[][] { { 2, 5000 }, { 2, 50000 }, { 5, 50000 }, { 10, 50000 }, { 20, 50000 }, });
  }

  private ExecutorService executor;
  private int threads;
  private int loops;
  private static RedissonClient regisClient;

  public AVMRedisMTPTest(int threads, int loops) {
    this.threads = threads;
    executor = Executors.newFixedThreadPool(threads);
    this.loops = loops;
  }

  @BeforeClass
  public static void setUpClassBefore() throws Exception {
    Config config = new Config();
    config.useSingleServer().setAddress("localhost:6379");
    regisClient = Redisson.create(config);
  }

  @AfterClass
  public static void tearDownClassAfter() throws Exception {
    regisClient.shutdown();
  }

  @Test
  public void lockUnlockRLock() throws Exception {
    for (int i = 0; i < threads; i++) {
      final String t = "T_" + i;

      Runnable worker = new Runnable() {

        @Override
        public void run() {
          for (int j = 0; j < loops; j++) {
            RLock lock = regisClient.getLock("RLOCK_" + j);
            lock.lock();
            try {
              RBucket<String> bucket = regisClient.getBucket("RBUCKET_" + j);
              bucket.set("TEST", 30, TimeUnit.SECONDS);
              RSemaphore semaphore = regisClient.getSemaphore("SEMAPHORE_" + j);
              semaphore.release();
              try {
                semaphore.acquire();
              } catch (InterruptedException e) {
                e.printStackTrace();
              }
              semaphore.expire(30, TimeUnit.SECONDS);
            } finally {
              lock.unlock();
            }
          }
        }
      };
      executor.execute(worker);
    }
    executor.shutdown();
    executor.awaitTermination(threads * loops, TimeUnit.SECONDS);

  }

}
`

**Comment by alexs20** [comment]
Here is a example
`package tmp;

import java.util.Arrays;
import java.util.Collection;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.MethodSorters;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import org.redisson.Config;
import org.redisson.Redisson;
import org.redisson.RedissonClient;
import org.redisson.core.RBucket;
import org.redisson.core.RLock;
import org.redisson.core.RSemaphore;

@FixMethodOrder(MethodSorters.NAME_ASCENDING)
@RunWith(Parameterized.class)
public class RedisMTPTest {
  @Parameters
  public static Collection<Object[]> data() {

    return Arrays.asList(new Object[][] { { 2, 5000 }, { 2, 50000 }, { 5, 50000 }, { 10, 50000 }, { 20, 50000 }, });
  }

  private ExecutorService executor;
  private int threads;
  private int loops;
  private static RedissonClient regisClient;

  public RedisMTPTest(int threads, int loops) {
    this.threads = threads;
    executor = Executors.newFixedThreadPool(threads);
    this.loops = loops;
  }

  @BeforeClass
  public static void setUpClassBefore() throws Exception {
    Config config = new Config();
    config.useSingleServer().setAddress("localhost:6379");
    regisClient = Redisson.create(config);
  }

  @AfterClass
  public static void tearDownClassAfter() throws Exception {
    regisClient.shutdown();
  }

  @Test
  public void lockUnlockRLock() throws Exception {
    for (int i = 0; i < threads; i++) {
      final String t = "T_" + i;

      Runnable worker = new Runnable() {

        @Override
        public void run() {
          for (int j = 0; j < loops; j++) {
            RLock lock = regisClient.getLock("RLOCK_" + j);
            lock.lock();
            try {
              RBucket<String> bucket = regisClient.getBucket("RBUCKET_" + j);
              bucket.set("TEST", 30, TimeUnit.SECONDS);
              RSemaphore semaphore = regisClient.getSemaphore("SEMAPHORE_" + j);
              semaphore.release();
              try {
                semaphore.acquire();
              } catch (InterruptedException e) {
                e.printStackTrace();
              }
              semaphore.expire(30, TimeUnit.SECONDS);
            } finally {
              lock.unlock();
            }
          }
        }
      };
      executor.execute(worker);
    }
    executor.shutdown();
    executor.awaitTermination(threads * loops, TimeUnit.SECONDS);

  }

}
`

**Comment by mrniko** [comment]
@alexs20 Hi! thank you for your test. It's really helped me to reproduce this issue. I will work on fix, stay tuned, please.

**Comment by mrniko** [comment]
@kajelas @alexs20  Fixed. Please check

**Comment by alexs20** [comment]
@kajelas It was quick. ;) Thank you very much. I will test it soon.

**Comment by kajelas** [comment]
@mrniko @alexs20 
I am away from my test environment, but I have read your fix and believe your fix is correct:) 

It seems the async `RedisPubSubConnection.unsubscribe()` makes the channel in unsubscribing process can be used unexpectedly by another thread. If it happens in a heavily concurrent situation, it is possible that the channel of the second thread is unsubscribed soon after the subscription, which leads to the deadlock.

And your fix gives a proper sync when `PublishSubscribe` calls `MasterSlaveConnectionManager.unsubscribe()`, which ensures that the channel is properly removed when `PublishSubscribe.unsubscribe()` ends.

Have I explained the bug and the fix correctly? Also Hope the fix can go online soon:)

**Comment by alexs20** [comment]
Hi,
Tried to run different scenarios, with local and remote server. Looks like now everything works fine.
Thanks.

**Comment by mrniko** [comment]
> Have I explained the bug and the fix correctly? 

You were absolutely correct :)

**Comment by mrniko** [comment]
@alexs20 
> I will continue my massive testing to see if it is reproducible...

I'm waiting for your results

**Comment by alexs20** [comment]
> I'm waiting for your results
So far working well. Probably it was some network issues / glich. Going to test with multiple servers... will report back once it is done.

**Comment by alexs20** [comment]
Ok, the status of testing with servers cluster is sad. Threads stuck on Lock.lock() almost immediately. From 20 concurrent threads only one or two are managing to finish the loop. In my test no semaphores or buckets as before. only Lock.lock() and Lock.unlock(). And cluster consist from 3 servers running on the same machine.

**Comment by mrniko** [comment]
@alexs20 Where did it stuck at subscription ?

**Comment by mrniko** [comment]
@alexs20 I have just launched your test with cluster (6 nodes = 3 master and 3 slave) and it successfully passed.

**Comment by mrniko** [comment]
@alexs20 can you make sure that you're using version from master branch?

**Comment by alexs20** [comment]
Here is some details:
Cluster and test are running on Win10. On each server the folowing parameters were changed
port 7001 (or 7002 or 7003)
bind 192.168.0.102 127.0.0.1
cluster-enabled yes
cluster-node-timeout 5000
All servers showing the cluster state: OK.

Here is my JUnit:
package com.conceptwave.lock.tmp;

import java.util.Arrays;
import java.util.Collection;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.FixMethodOrder;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.MethodSorters;
import org.junit.runners.Parameterized;
import org.junit.runners.Parameterized.Parameters;
import org.redisson.Config;
import org.redisson.Redisson;
import org.redisson.RedissonClient;
import org.redisson.core.RLock;

@FixMethodOrder(MethodSorters.NAME_ASCENDING)
@RunWith(Parameterized.class)
public class RedisMTPMultiServer {
  @Parameters
  public static Collection<Object[]> data() {

    return Arrays.asList(new Object[][] { { 20, 500 }, });
  }

  private ExecutorService executor;
  private int threads;
  private int loops;
  private static RedissonClient regisClient;
  private AtomicInteger count;

  public RedisMTPMultiServer(int threads, int loops) {
    this.threads = threads;
    count = new AtomicInteger(threads);
    executor = Executors.newFixedThreadPool(threads);
    this.loops = loops;
  }

  @BeforeClass
  public static void setUpClassBefore() throws Exception {
    Config config = new Config();
     config.useClusterServers().addNodeAddress("192.168.0.102:7001").addNodeAddress("192.168.0.102:7002").addNodeAddress("192.168.0.102:7003");
    regisClient = Redisson.create(config);
  }

  @AfterClass
  public static void tearDownClassAfter() throws Exception {
    regisClient.shutdown();
  }

  @Test
  public void lockUnlockRLock() throws Exception {
    for (int i = 0; i < threads; i++) {
      Runnable worker = new Runnable() {

        @Override
        public void run() {
          for (int j = 0; j < loops; j++) {
            RLock lock = regisClient.getLock("RLOCK_" + j);
            lock.lock();
            try {
              //dooing something
            } finally {
              lock.unlock();
            }
          }
          System.out.println(count.decrementAndGet() + " threads left");
        }
      };
      executor.execute(worker);
    }
    executor.shutdown();
    executor.awaitTermination(Long.MAX_VALUE, TimeUnit.MILLISECONDS);

  }

}

And here is stack-trace of stuck thread (all threads staying on the same line):
Unsafe.park(boolean, long) line: not available [native method]	
LockSupport.park(Object) line: not available	
CountDownLatch$Sync(AbstractQueuedSynchronizer).parkAndCheckInterrupt() line: not available	
CountDownLatch$Sync(AbstractQueuedSynchronizer).doAcquireSharedInterruptibly(int) line: not available	
CountDownLatch$Sync(AbstractQueuedSynchronizer).acquireSharedInterruptibly(int) line: not available	
CountDownLatch.await() line: not available	
CommandSyncService(CommandAsyncService).get(Future<V>) line: 95	
RedissonLock(RedissonObject).get(Future<V>) line: 55	
RedissonLock.lockInterruptibly(long, TimeUnit) line: 121	
RedissonLock.lockInterruptibly() line: 108	
RedissonLock.lock() line: 90	
RedisMTPMultiServer$1.run() line: 68	
ThreadPoolExecutor.runWorker(ThreadPoolExecutor$Worker) line: not available	
ThreadPoolExecutor$Worker.run() line: not available	
Thread.run() line: not available	


Here is the output to console:

07:52:22.011 [main-null] INFO  org.redisson.Version - Redisson 2.2.16.SNAPSHOT
07:52:22.576 [nioEventLoopGroup-2-1-null] INFO  o.r.c.p.MasterConnectionPool - 5 connections initialized for /192.168.0.102:7002
07:52:22.582 [main-null] INFO  o.r.c.ClusterConnectionManager - master: //192.168.0.102:7002 added for slot ranges: [[5461-10922]]
07:52:22.591 [nioEventLoopGroup-2-8-null] INFO  o.r.c.p.MasterConnectionPool - 5 connections initialized for /192.168.0.102:7001
07:52:22.593 [nioEventLoopGroup-2-8-null] INFO  o.r.c.ClusterConnectionManager - master: //192.168.0.102:7001 added for slot ranges: [[0-5460]]
07:52:22.601 [nioEventLoopGroup-2-6-null] INFO  o.r.c.p.MasterConnectionPool - 5 connections initialized for /192.168.0.102:7003
07:52:22.601 [nioEventLoopGroup-2-6-null] INFO  o.r.c.ClusterConnectionManager - master: //192.168.0.102:7003 added for slot ranges: [[10923-16383]]
19 threads left


And about latest code - I did git clone last day and compiled snapshot jar.


BTW, you are starting to use Java 1.8 syntax and this prevent me from compiling it with JDK 1.7 as is.
I had to strip out all test (uses lambda) and added a few "final" statements in main code.
I just want to warn you that most of enterprise sectors using Java 7 (like our company) and If you will limit your library to Java 8 only it will kill the interest to the project, and as result project itself.

**Comment by alexs20** [comment]
A few cents,
Just tested with cluster on linux box (Mint 18). The same result :(

**Comment by mrniko** [comment]
> BTW, you are starting to use Java 1.8 syntax and this prevent me from compiling it with JDK 1.7 as is.
master branch is still 1.6 compatible.

**Comment by mrniko** [comment]
I have just run your latest test and it passed too. I'm using win7 to run Redisson and VirtualBox with Ubuntu server 14 to run redis cluster.

**Comment by mrniko** [comment]
> I just want to warn you that most of enterprise sectors using Java 7 (like our company) and If you will limit your library to Java 8 only it will kill the interest to the project, and as result project itself.

Don't worry about it. 2.x.x will have a support

**Comment by alexs20** [comment]
OK, Looks like I have found when it does work and when doesnt.
When cluster consist 3 masters + 3 slaves then everything is fine
but when only 3 masters (no slaves configured at all ... for example I do not want persistence but need high scalability) then test fail.
Gotcha! :)

**Comment by alexs20** [comment]
BTW, master branch contains unsolvable compilation problem:
org/redisson/spring/cache/RedissonCache.java line 155: Unresolved type: ValueRetrievalException

**Comment by mrniko** [comment]
Thank you! will test it now! 

> unsolvable compilation problem

Spring 4.3.0 has out so you need to rebuild your IDE project files.

**Comment by mrniko** [comment]
@alexs20 Fixed!

**Comment by alexs20** [comment]
> Fixed!

Thanks! Testing...

**Comment by alexs20** [comment]
So far I do not see critical problems. At least in simple tests. More complex solutions that we a trying to implement using different redisson objects still showing some wired behavior when sometimes it takes too long to obtain the lock (looks like stuck, but then it continue to run) but I need more time to investigate that and I am not sure where exactly the problem is. So I will report once I found something...
Thanks.

**Comment by mrniko** [comment]
Thank you! If you found a new issue, please report it
